docs/ROADMAP.md

```diff
@@ -0,0 +1,382 @@
+API* section, since they usually only apply to a single module or type. But the
+biggest changes are explained in detail here.
+
+### Split per HTTP version
+
+The existing `Connection` types, both for the client and server, abstract over
+HTTP version by requiring a generic `AsyncRead + AsyncWrite` transport type.
+But as we figure out HTTP/3, that needs to change. So to prepare now, the
+`Connection` types will be split up.
+
+For example, there will now be `hyper::server::conn::http1::Connection` and
+`hyper::server::conn::http2::Connection` types.
+
+These specific types will still have a very similar looking API that, as the
+VISION describes, provides **Correct** connection management as it pertains to
+HTTP.
+
+There will be still be a type to wrap the different versions. It will no longer
+be generic over the transport type, to prepare for being able to wrap HTTP/3
+connections. Exactly how it will wrap, either by using internal trait objects,
+or an `enum Either` style, or using a `trait Connection` that each type
+implements, is something to be determined. It's likely that this "auto" type
+will start in `hyper-util`.
+
+### Focus on the `Connection` level
+
+As mentioned in the *Known Issues*, the higher-level `Client` and `Server` have
+stability and complexity problems. Therefore, for hyper 1.0, the main API will
+focus on the "lower-level" connection types. The `Client` and `Server` helpers
+will be moved to `hyper-util`.
+
+## Public API
+
+### body
+
+The `Body` struct is removed. Its internal "variants" are [separated into
+distinct types](https://github.com/hyperium/hyper/issues/2345), and can start
+in either `hyper-util` or `http-body-util`.
+
+The exported trait `HttpBody` is renamed to `Body`.
+
+A single `Body` implementation in `hyper` is the one provided by receiving
+client responses and server requests. It has the name `Streaming`.
+
+> **Unresolved**: Other names can be considered during implementation. Another
+> option is to not publicly name the implementation, but return `Response<impl
+Body>`s.
+
+The `Body` trait will be experimented on to see about making it possible to
+return more frame types beyonds just data and trailers.
+
+> **Unresolved**: What exactly this looks like will only be known after
+> experimentation.
+
+### client
+
+The high-level `hyper::Client` will be removed, along with the
+`hyper::client::connect` module. They will be explored more in `hyper-util`.
+
+As described in *Design*, the `client::conn` module will gain `http1` and
+`http2` sub-modules, providing per-version `SendRequest`, `Connection`, and
+`Builder` structs. An `auto` version can be explored in `hyper-util`.
+
+### error
+
+The `hyper::Error` struct remains in place.
+
+All errors returned from `Error::source()` are made opaque. They are wrapped an
+internal `Opaque` newtype that still allows printing, but prevents downcasting
+to the internal dependency.
```

Comment:
Ruling out downcasting but allowing printing means that our internal dependencies will still leak; applications will end up breaking at runtime due to string matching failures rather than downcasting failures.

The only way we can move breakage from runtime to compile time is to make our internal dependencies part of the public API, therefore subjecting all of hyper to their semver decisions. Not good.

I try to avoid them wherever possible, but perhaps this is a case where feature flags could let users opt in to having dependency error types in the public API at the cost of also having to explicitly opt into non-semver-conforming behavior from hyper with respect to those dependencies.

In other words, if you don't opt in, the best you can do is string matching or downcasting, but you won't have any surprise requirement to update your own `h2` dependency by a major version if hyper bumps its own dependency in a minor release. If you _do_ opt in, then `Error::h2_reason()` becomes `pub` instead of `pub(super)`, but you might have to be more strict about your hyper upgrades.

This is... quite a mess, but I think it's also less messy than depending on stability of `Display` impls for downstream metrics and reporting.

Reply:
As I mentioned in the other comment, I believe that a small group of errors should be handled programmatically, and the rest should just be printed to a human. I like to believe that most users know that format string matching is very fragile, and hopefully not depended on for much.

I think downcasting an error can be useful, but too much coupling to the specific error types that exist in a stack can make a user's software too fragile. They should be able to inspect errors we purposefully expose, but otherwise we should be allowed to change internal dependencies.

I know @nox was in favor of this change, he might have some more thoughts.

Reply:
I disagree that we should care about whether we could break things through `Display` and `Debug`, but I do think we should rewrap errors in Hyper so that we keep dependencies truly private.

I don't want `Opaque`, I want a wrapper of `h2::Error`, so that we can still downcast *and* not suffer from h2 bumps *and* not even need to depend on h2 at all.

The only reason some arbitrary crates depend on h2 at work is to be able to do `err.downcast::<h2::Error>()`, if that error was wrapped in Hyper, then I would do `err.downcast::<hyper::H2Error>()` and I wouldn't ever depend on h2 directly.

Reply:
> I want a wrapper of `h2::Error`, so that we can still downcast and not suffer from h2 bumps and not even need to depend on h2 at all.

Just to make sure I understand: you want a type in `hyper` that mimics `h2::Error` but does not reëxport it? Would that be substantially different from h2-specific variants on the proposed `Code` type?

Reply:
I feel that the `Code` proposal in this document would address that.

Reply:
I guess it does indeed.